### PR TITLE
Add GitHub Copilot native

### PR DESCRIPTION
**Pull request by diopresi**
Add GitHub Copilot native integration with Backward compatibility

### CODE REVIEW

**Comment by diopresi** [comment]
Hi @saoudrizwan, When you get a chance, could you take a look? If everything looks good, feel free to merge it.

Thanks!

**Comment by diopresi** [comment]
Relates to Add GitHub Copilot provider https://github.com/cline/cline/discussions/660

**Comment by twisteddebian** [comment]
Hi there, I tried your implementation, but I’m encountering an issue where Claude seems to get stuck in a loop. The first API message displays correctly, Claude summarizes a first overhead, but then it proceeds to read a few files and eventually hangs on a specific file, repeatedly making the API request with the message: "Claude, read this file:". A continuous loop, which I'm only experiencing with the Copilot integration, Anthropic API works fine. What may cause this?

Have you experienced anything similar during testing of your updates? I've tried multiple Workspaces, VS Code is on version 1.95.3. Looking forward to this game-changing implementation using Github Copilot.

File: src/shared/api.ts [hunk 1]
```diff
@@ -323,3 +324,83 @@ export const openAiNativeModels = {
...
+	"claude-3-5-sonnet": {
+		maxTokens: 194827,
+		contextWindow: 194827,
+		supportsImages: false,
+		supportsComputerUse: false,
+		supportsPromptCache: false,
+		inputPrice: 0.0,
+		outputPrice: 0.0
+	}
+} as const satisfies Record<string, ModelInfo>
```
Comment (lloydchang) [inline src/shared/api.ts:406]: **Question:** Where do the numbers in these lines, such as:
```TypeScript
	"gpt-4o": {
		maxTokens: 123734
```
originate?

---

The URL you referenced:
https://code.visualstudio.com/api/extension-guides/language-model#related-content
provides a different set of numbers.

---

For example:
https://code.visualstudio.com/api/extension-guides/language-model#choosing-the-appropriate-model
says:
> **Note:** The recommended GPT-4o model has a limit of `64K tokens`. The returned model object from the `selectChatModels` call has a `maxInputTokens` attribute that shows the token limit. These limits will be expanded as we learn more about how extensions are using the language models.

---

Specifically, `123734` is different than `64K`.

---

**To recap:**

```TypeScript
	"gpt-4o": {
		maxTokens: 123734
```
is different than
> The recommended GPT-4o model has a limit of `64K tokens`.

**Review by lloydchang — CHANGES_REQUESTED** [review]

**Comment by lloydchang** [comment]
> Hi there, I tried your implementation, but I’m encountering an issue where Claude seems to get stuck in a loop. The first API message displays correctly, Claude summarizes a first overhead, but then it proceeds to read a few files and eventually hangs on a specific file, repeatedly making the API request with the message: "Claude, read this file:". A continuous loop, which I'm only experiencing with the Copilot integration, Anthropic API works fine. What may cause this?
> 
> Have you experienced anything similar during testing of your updates? I've tried multiple Workspaces, VS Code is on version 1.95.3. Looking forward to this game-changing implementation using Github Copilot.

Ditto.

**Comment by Mithadon** [comment]
Even if it gets rate limited quickly, I'd love to be able to save some API money by switching to Copilot now and then. Hope this gets added

**Comment by lloydchang** [comment]
> Even if it gets rate limited quickly, I'd love to be able to save some API money by switching to Copilot now and then. Hope this gets added

In the meantime, why not use Copilot directly?

The implementation in this pull request is based on https://code.visualstudio.com/api/extension-guides/language-model#choosing-the-appropriate-model, which isn't very different from what Copilot already offers.

My understanding is that this approach aims to extend Visual Studio by using the Language Model AI, which requires Copilot, rather than directly utilizing the AI models.

It doesn't support images, computer use, nor prompt caches.

**Comment by lloydchang** [comment]
Thanks for sharing the proposed changes.

**Another Question:** Is this pull request expecting https://code.visualstudio.com/api/extension-guides/language-model to _eventually(?)_ support tools / function calling?

**Comment by Mithadon** [comment]
> > Even if it gets rate limited quickly, I'd love to be able to save some API money by switching to Copilot now and then. Hope this gets added
> 
> In the meantime, why not use Copilot directly?
> 
> The implementation in this pull request is based on https://code.visualstudio.com/api/extension-guides/language-model#choosing-the-appropriate-model, which isn't very different from what Copilot already offers.
> 
> My understanding is that this approach aims to extend Visual Studio by using the Language Model AI, which requires Copilot, rather than directly utilizing the AI models.
> 
> It doesn't support images, computer use, nor prompt caches.

In that case, I misunderstood. Yes, right now I use a combination of copilot and Cline simultaneously. Cline's tool use is just too good.

**Comment by lloydchang** [comment]
>  I use a combination of copilot

Have you tried GitHub Copilot Workspace in Technical Preview?

---

Its Brainstorm and Plan features are comparable to Cline's Task, but they operate at different levels of abstraction:

**Abstraction Layers in GitHub Copilot Workspace:**
- Brainstorm
  - Task (GitHub Copilot Workspace's Task is similar to Cline's Task)
    - Plan
      - File
        - Item

---

**References:**

GitHub Copilot Workspace integrates with the web, GitHub Codespaces, and Visual Studio Code:

1. https://github.blog/news-insights/product-news/github-copilot-workspace/
2. https://githubnext.com/projects/copilot-workspace  
3. https://copilot-workspace.githubnext.com/  
4. https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-workspace  
5. https://github.com/githubnext/copilot-workspace-user-manual/tree/main  
6. https://visualstudiomagazine.com/Articles/2024/10/24/Copilot-Workspace.aspx

---

> Cline's tool use is just too good.

Are you referring to Anthropic's [Computer Use](https://docs.anthropic.com/en/docs/build-with-claude/computer-use) and Claude 3.5 Sonnet's [Agentic Coding](https://www-cdn.anthropic.com/fed9cc193a14b84131812372d8d5857f8f304c52/Model_Card_Claude_3_Addendum.pdf)?

---

> save some API money

Compared to Anthropic as an API Provider, if you're using OpenRouter,  there is an [additional 5% plus .35 cent per transaction](https://openrouter.ai/terms).

Alternatively in Cline, you can switch to OpenRouter's free models, Gemini's free models, or your computer's local models served by LM Studio or Ollama.

**Comment by saoudrizwan** [comment]
Thank you @diopresi for the contribution! Note to reviewers: we need to compare this branch to https://github.com/cline/cline/pull/970 and see how we can go about merging both in + resolve conflicts.

**Comment by diopresi** [comment]
Thank you all for reviewing this and providing your feedback!

Unfortunately, this implementation falls short compared to other LLM providers. The main limitation is that GitHub currently does not offer a way to add or modify the System Prompt. As a result, conversations with the GitHub provider are not handled effectively, since GitHub injects its own System Prompt before it reaches the underlying LLM.

Given that GitHub is unlikely to provide a solution to unlock this capability anytime soon, I’ve decided to close this PR for now.

**Comment by Bigsy** [comment]
> Thank you all for reviewing this and providing your feedback!
> 
> Unfortunately, this implementation falls short compared to other LLM providers. The main limitation is that GitHub currently does not offer a way to add or modify the System Prompt. As a result, conversations with the GitHub provider are not handled effectively, since GitHub injects its own System Prompt before it reaches the underlying LLM.
> 
> Given that GitHub is unlikely to provide a solution to unlock this capability anytime soon, I’ve decided to close this PR for now.

Roocline have added a version of this, works great for me, built a decent size chrome extension nicely, didn't even feel like I was missing full fat sonnet and better than deepseek3
